IllegalStateException: Cannot perform this operation because the connection pool has been closed.

Crash report from multiple users:

- Sent: 27.12.2019 - 30.12.2019
- App versions: [v.1.41.0](https://github.com/EventFahrplan/EventFahrplan/releases/tag/v.1.41.0) - [v.1.41.4](https://github.com/EventFahrplan/EventFahrplan/releases/tag/v.1.41.4)
- Android versions: 5.0.1, 6.0.1, 7.0, 7.1.1, 8.1.0, 9, 10
- Device models: XT1676, Pixel XL, Aquaris X Pro, A0001, SM-G955F, Nexus 7, ONEPLUS A5010, Pixel 3a, Moto G Play, SM-G920F, ...

# Stacktrace
``` java
java.lang.RuntimeException: Unable to resume activity {info.metadude.android.congress.schedule/
    nerd.tuxmobil.fahrplan.congress.favorites.StarredListActivity}: 
    java.lang.IllegalStateException: 
    Cannot perform this operation because the connection pool has been closed.
at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3823)
at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3855)
at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:51)
at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:145)
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1817)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:193)
at android.app.ActivityThread.main(ActivityThread.java:6723)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:495)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)

Caused by: java.lang.IllegalStateException: 
    Cannot perform this operation because the connection pool has been closed.
at android.database.sqlite.SQLiteConnectionPool.throwIfClosedLocked(SQLiteConnectionPool.java:1069)
at android.database.sqlite.SQLiteConnectionPool.waitForConnection(SQLiteConnectionPool.java:683)
at android.database.sqlite.SQLiteConnectionPool.acquireConnection(SQLiteConnectionPool.java:378)
at android.database.sqlite.SQLiteSession.acquireConnection(SQLiteSession.java:894)
at android.database.sqlite.SQLiteSession.executeForCursorWindow(SQLiteSession.java:834)
at android.database.sqlite.SQLiteQuery.fillWindow(SQLiteQuery.java:62)
at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:149)
at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:137)
at info.metadude.android.eventfahrplan.database.repositories.LecturesDatabaseRepository.query(LecturesDatabaseRepository.kt:78)
at info.metadude.android.eventfahrplan.database.repositories.LecturesDatabaseRepository.queryLecturesOrderedByDateUtc(LecturesDatabaseRepository.kt:45)
at nerd.tuxmobil.fahrplan.congress.repositories.AppRepository.readLecturesOrderedByDateUtc(AppRepository.kt:288)
at nerd.tuxmobil.fahrplan.congress.repositories.AppRepository.loadLecturesForDayIndex(AppRepository.kt:230)
at nerd.tuxmobil.fahrplan.congress.repositories.AppRepository.loadLecturesForAllDays(AppRepository.kt:219)
at nerd.tuxmobil.fahrplan.congress.utils.FahrplanMisc.getStarredLectures(FahrplanMisc.java:200)
at nerd.tuxmobil.fahrplan.congress.favorites.StarredListFragment.initStarredList(StarredListFragment.java:131)
at nerd.tuxmobil.fahrplan.congress.favorites.StarredListFragment.onResume(StarredListFragment.java:125)
at android.support.v4.app.Fragment.performResume(Fragment.java:2498)
at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1501)
at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1784)
at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1852)
at android.support.v4.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3269)
at android.support.v4.app.FragmentManagerImpl.dispatchResume(FragmentManager.java:3241)
at android.support.v4.app.FragmentController.dispatchResume(FragmentController.java:223)
at android.support.v4.app.FragmentActivity.onResumeFragments(FragmentActivity.java:538)
at android.support.v4.app.FragmentActivity.onPostResume(FragmentActivity.java:527)
at android.support.v7.app.AppCompatActivity.onPostResume(AppCompatActivity.java:172)
at android.app.Activity.performResume(Activity.java:7325)
at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3815)
... 11 more
```

# Log from one user
``` java
1 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag=W/"1ea297f6822f758b1a2fb68482f6ab101dcfa592da170ac11737199a4ae9d980", numDays=4, subtitle=, title=36th Chaos Communication Congress, version=find; lounges 000; chaos-west 1.0; open-infra 13; wikipaka 0.7; chaoszone 0.7; komona 0.8; sendezentrum 2.2; lightning Samarium; art-play 0.24; cdc 3; wiki 2019-12-27 11:20)
2 D: viewDay(false)
3 D: fetch was pending, restart
4 D: MyApp.task_running = FETCH
5 D: lectureId = null
6 D: MyApp.lectureList contains 274 items.
7 D: onResume
8 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag=W/"1ea297f6822f758b1a2fb68482f6ab101dcfa592da170ac11737199a4ae9d980", numDays=4, subtitle=, title=36th Chaos Communication Congress, version=find; lounges 000; chaos-west 1.0; open-infra 13; wikipaka 0.7; chaoszone 0.7; komona 0.8; sendezentrum 2.2; lightning Samarium; art-play 0.24; cdc 3; wiki 2019-12-27 11:20)
9 D: viewDay(false)
10 D: meta.getNumDays() = 4
11 D: MyApp.task_running = NONE
12 D: lectureId = null
13 D: MyApp.lectureList contains 274 items.
14 D: onResume
15 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag=W/"1ea297f6822f758b1a2fb68482f6ab101dcfa592da170ac11737199a4ae9d980", numDays=4, subtitle=, title=36th Chaos Communication Congress, version=find; lounges 000; chaos-west 1.0; open-infra 13; wikipaka 0.7; chaoszone 0.7; komona 0.8; sendezentrum 2.2; lightning Samarium; art-play 0.24; cdc 3; wiki 2019-12-27 11:20)
16 D: viewDay(false)
17 D: meta.getNumDays() = 4
18 D: MyApp.task_running = NONE
19 D: lectureId = null
20 D: MyApp.lectureList contains 274 items.
21 D: onResume
22 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag=W/"1ea297f6822f758b1a2fb68482f6ab101dcfa592da170ac11737199a4ae9d980", numDays=4, subtitle=, title=36th Chaos Communication Congress, version=find; lounges 000; chaos-west 1.0; open-infra 13; wikipaka 0.7; chaoszone 0.7; komona 0.8; sendezentrum 2.2; lightning Samarium; art-play 0.24; cdc 3; wiki 2019-12-27 11:20)
23 D: viewDay(false)
24 D: meta.getNumDays() = 4
25 D: MyApp.task_running = NONE
26 D: lectureId = null
27 D: MyApp.lectureList contains 274 items.
28 D: onResume
29 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag=W/"1ea297f6822f758b1a2fb68482f6ab101dcfa592da170ac11737199a4ae9d980", numDays=4, subtitle=, title=36th Chaos Communication Congress, version=find; lounges 000; chaos-west 1.0; open-infra 13; wikipaka 0.7; chaoszone 0.7; komona 0.8; sendezentrum 2.2; lightning Samarium; art-play 0.24; cdc 3; wiki 2019-12-27 11:20)
30 D: viewDay(false)
31 D: meta.getNumDays() = 4
32 D: MyApp.task_running = NONE
33 D: lectureId = null
34 D: MyApp.lectureList contains 274 items.
35 D: onResume
36 D: MyApp.meta = Meta(dayChangeHour=0, dayChangeMinute=0, eTag=W/"1ea297f6822f758b1a2fb68482f6ab101dcfa592da170ac11737199a4ae9d980", numDays=4, subtitle=, title=36th Chaos Communication Congress, version=find; lounges 000; chaos-west 1.0; open-infra 13; wikipaka 0.7; chaoszone 0.7; komona 0.8; sendezentrum 2.2; lightning Samarium; art-play 0.24; cdc 3; wiki 2019-12-27 11:20)
37 D: viewDay(false)
38 D: meta.getNumDays() = 4
39 D: MyApp.task_running = NONE
40 D: lectureId = null
41 D: MyApp.lectureList contains 274 items.
```

# User reports
- User 01:
  > Eben habe ich mein Handy mit offener app gesperrt. Dann habe ich entsperrt und dann direkt meine Favoritenliste geöffnet.

- User 02:
  > Die App ist beim Wechsel in den Landscape abgestürzt.

- User 03:
  > Hatte das Telefon in der Tasche, es hat vibriert wegen einem Schedule update, hab sie geöffnet, auf später geklickt und drauf ist sie abgestürzt

- User 04:
  > nicht sicher wie man das genau reproduziert. oft wenn man sie wieder in den Vordergrund holt und Updates rein kamen.

- User 05:
  > Schnell hin und her gedreht

- User 06:
  > geöffnet, Aktualisierungsnachricht bestätigt, und dann sofort auf Favoriten-Page gewechselt.

# Related
- [Issue #229: SQLiteDatabaseLockedException: database is locked](https://github.com/EventFahrplan/EventFahrplan/issues/229)

--- a/app/src/main/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmList.java
+++ b/app/src/main/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmList.java
@@ -63,7 +63,6 @@ public class AlarmList extends ActionBarListActivity {
                     null, null, null, Columns.TIME);
         } catch (SQLiteException e) {
             e.printStackTrace();
-            db.close();
             return;
         }
         startManagingCursor(cursor);
@@ -76,12 +75,6 @@ public class AlarmList extends ActionBarListActivity {
         registerForContextMenu(getListView());
 
         setResult(RESULT_CANCELED);
-    }
-
-    @Override
-    protected void onDestroy() {
-        super.onDestroy();
-        db.close();
     }
 
     @Override

--- a/app/src/main/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepository.kt
+++ b/app/src/main/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepository.kt
@@ -258,9 +258,8 @@ object AppRepository {
         alarmsDatabaseRepository.query(eventId).toAlarmsAppModel()
     }
 
-    @JvmOverloads
-    fun deleteAlarmForAlarmId(alarmId: Int, closeSQLiteOpenHelper: Boolean = true) =
-            alarmsDatabaseRepository.deleteForAlarmId(alarmId, closeSQLiteOpenHelper)
+    fun deleteAlarmForAlarmId(alarmId: Int) =
+            alarmsDatabaseRepository.deleteForAlarmId(alarmId)
 
     fun deleteAlarmForEventId(eventId: String) =
             alarmsDatabaseRepository.deleteForEventId(eventId)

--- a/database/src/main/java/info/metadude/android/eventfahrplan/database/extensions/SQLiteDatabaseExtensions.kt
+++ b/database/src/main/java/info/metadude/android/eventfahrplan/database/extensions/SQLiteDatabaseExtensions.kt
@@ -35,24 +35,20 @@ fun SQLiteDatabase.delete(tableName: String, columnName: String? = null, columnV
 }
 
 /**
- * Executes the delete [query] within a transaction and finally closes the database.
+ * Executes the delete [query] within a transaction.
  */
 internal fun SQLiteDatabase.delete(query: SQLiteDatabase.() -> Int) =
-        use {
-            it.transaction {
-                query()
-            }
+        transaction {
+            query()
         }
 
 /**
- * Executes the [delete] and the [insert] queries within a transaction and finally closes the database.
+ * Executes the [delete] and the [insert] queries within a transaction.
  */
 internal fun SQLiteDatabase.upsert(delete: SQLiteDatabase.() -> Int, insert: SQLiteDatabase.() -> Long) =
-        use {
-            it.transaction {
-                delete()
-                insert()
-            }
+        transaction {
+            delete()
+            insert()
         }
 
 /**

--- a/database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/AlarmsDatabaseRepository.kt
+++ b/database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/AlarmsDatabaseRepository.kt
@@ -23,7 +23,6 @@ class AlarmsDatabaseRepository(
         }, {
             insert(AlarmsTable.NAME, values)
         })
-        close()
     }
 
     fun query(): List<Alarm> = query {
@@ -44,15 +43,11 @@ class AlarmsDatabaseRepository(
         } catch (e: SQLiteException) {
             e.printStackTrace()
             Log.e(javaClass.name, "Failure on alarm query.")
-            database.close()
-            sqLiteOpenHelper.close()
             return alarms.toList()
         }
 
         if (cursor.count == 0) {
             cursor.close()
-            database.close()
-            sqLiteOpenHelper.close()
             Log.d(javaClass.name, "No alarms found.")
             return alarms.toList()
         }
@@ -70,13 +65,11 @@ class AlarmsDatabaseRepository(
             cursor.moveToNext()
         }
         cursor.close()
-        database.close()
-        sqLiteOpenHelper.close()
 
         return alarms.toList()
     }
 
-    fun deleteForAlarmId(alarmId: Int, closeSQLiteOpenHelper: Boolean) = delete(closeSQLiteOpenHelper) {
+    fun deleteForAlarmId(alarmId: Int) = delete {
         delete(AlarmsTable.NAME, ID, "$alarmId")
     }
 
@@ -84,12 +77,9 @@ class AlarmsDatabaseRepository(
         delete(AlarmsTable.NAME, EVENT_ID, eventId)
     }
 
-    private fun delete(closeSQLiteOpenHelper: Boolean = true, query: SQLiteDatabase.() -> Int) =
+    private fun delete(query: SQLiteDatabase.() -> Int) =
             with(sqLiteOpenHelper) {
                 writableDatabase.delete(query)
-                if (closeSQLiteOpenHelper) {
-                    close()
-                }
             }
 
 }

--- a/database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/HighlightsDatabaseRepository.kt
+++ b/database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/HighlightsDatabaseRepository.kt
@@ -23,7 +23,6 @@ class HighlightsDatabaseRepository(
         }, {
             insert(HighlightsTable.NAME, values)
         })
-        close()
     }
 
     fun query(): List<Highlight> {
@@ -34,8 +33,6 @@ class HighlightsDatabaseRepository(
             cursor = database.read(HighlightsTable.NAME)
         } catch (e: SQLiteException) {
             e.printStackTrace()
-            database.close()
-            sqLiteOpenHelper.close()
             return highlights.toList()
         }
 
@@ -50,8 +47,6 @@ class HighlightsDatabaseRepository(
             cursor.moveToNext()
         }
         cursor.close()
-        database.close()
-        sqLiteOpenHelper.close()
 
         return highlights.toList()
     }
@@ -65,8 +60,6 @@ class HighlightsDatabaseRepository(
                 selectionArgs = arrayOf(eventId.toString())
             )
         } catch (e: SQLiteException) {
-            database.close()
-            sqLiteOpenHelper.close()
             return null
         }
 
@@ -80,8 +73,6 @@ class HighlightsDatabaseRepository(
             }
         }
 
-        database.close()
-        sqLiteOpenHelper.close()
         return highlight
     }
 

--- a/database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/LecturesDatabaseRepository.kt
+++ b/database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/LecturesDatabaseRepository.kt
@@ -9,6 +9,7 @@ import info.metadude.android.eventfahrplan.database.contract.FahrplanContract.Le
 import info.metadude.android.eventfahrplan.database.contract.FahrplanContract.LecturesTable.Columns.*
 import info.metadude.android.eventfahrplan.database.contract.FahrplanContract.LecturesTable.Values.REC_OPT_OUT_OFF
 import info.metadude.android.eventfahrplan.database.extensions.*
+import info.metadude.android.eventfahrplan.database.extensions.transaction
 import info.metadude.android.eventfahrplan.database.models.Lecture
 import info.metadude.android.eventfahrplan.database.sqliteopenhelper.LecturesDBOpenHelper
 
@@ -20,15 +21,12 @@ class LecturesDatabaseRepository(
 ) {
 
     fun insert(list: List<ContentValues>) = with(sqLiteOpenHelper) {
-        writableDatabase.use {
-            it.transaction {
-                delete(LecturesTable.NAME)
-                list.forEach { contentValues ->
-                    insert(LecturesTable.NAME, contentValues)
-                }
+        writableDatabase.transaction {
+            delete(LecturesTable.NAME)
+            list.forEach { contentValues ->
+                insert(LecturesTable.NAME, contentValues)
             }
         }
-        close()
     }
 
     fun queryLectureByLectureId(lectureId: String): Lecture {
@@ -79,15 +77,11 @@ class LecturesDatabaseRepository(
             cursor = query()
         } catch (e: SQLiteException) {
             e.printStackTrace()
-            close()
-            sqLiteOpenHelper.close()
             return lectures.toList()
         }
 
         if (cursor.count == 0) {
             cursor.close()
-            close()
-            sqLiteOpenHelper.close()
             return lectures.toList()
         }
 
@@ -139,8 +133,6 @@ class LecturesDatabaseRepository(
             cursor.moveToNext()
         }
         cursor.close()
-        close()
-        sqLiteOpenHelper.close()
         return lectures.toList()
     }
 

--- a/database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/MetaDatabaseRepository.kt
+++ b/database/src/main/java/info/metadude/android/eventfahrplan/database/repositories/MetaDatabaseRepository.kt
@@ -27,7 +27,6 @@ class MetaDatabaseRepository(
         }, {
             insert(MetasTable.NAME, values)
         })
-        close()
     }
 
     fun query(): Meta {
@@ -38,8 +37,6 @@ class MetaDatabaseRepository(
             cursor = database.read(MetasTable.NAME)
         } catch (e: SQLiteException) {
             e.printStackTrace()
-            database.close()
-            sqLiteOpenHelper.close()
             return Meta()
         }
 
@@ -61,8 +58,6 @@ class MetaDatabaseRepository(
         Log.d(javaClass.name, "query(): $meta")
 
         cursor.close()
-        database.close()
-        sqLiteOpenHelper.close()
 
         return meta
     }

